Exercises/NoModules/Chapter 15/Soln15_04/Zoo.cpp

```diff
@@ -0,0 +1,58 @@
+// Implementations of the Zoo class that stores pointers to Animals
+#include "Zoo.h"
+#include "Animals.h"
+#include <iostream>
+
+// Constructor from a vector of animals
+Zoo::Zoo(const std::vector<AnimalPtr>& animals) 
+  : m_animals{ animals } 
+{}
+
+// Add an animal to the zoo
+void Zoo::addAnimal(AnimalPtr animal)
+{
+  m_animals.push_back(animal);
+}
+
+// Output the animals and the sound they make
+void Zoo::showAnimals() const
+{
+  for (const auto& animal : m_animals)
+  {
+    std::cout << animal->who() << ' ' << animal->sound() << std::endl;
+  }
+}
+
+// Collect all Sheep in the Zoo using dynamic cast (the recommended way)
+std::vector<Sheep*> Zoo::herd() const
+{
+  std::vector<Sheep*> sheep;
+  for (auto animal : m_animals)
+  {
+    auto* casted = dynamic_cast<Sheep*>(animal.get());
+    if (casted)
+    {
+      sheep.push_back(casted);
+    }
+  }
+  return sheep;
+}
+
+/*
+// Collect all Sheep in the Zoo using the typeid() operator and a static cast
+// (for the sake of the exercise only)
```

Comment:
i wouldn't show this, but then again it's being asked in the exercise...

Reply:
They learned about typeid() in this chapter, so it's worth the exercise.